src/daemon/daemon.cpp

```diff
@@ -1544,119 +1722,72 @@ try // clang-format on
+                                                                 const SnapshotPick& snapshot_pick) {
+        for (const auto& snapshot_name : snapshot_pick.pick)
         {
-            try
-            {
-                auto vm_image_info = config->image_hosts.back()->info_for_full_hash(vm_image.id);
-                original_release = vm_image_info.release_title.toStdString();
-            }
-            catch (const std::exception& e)
-            {
-                mpl::log(mpl::Level::warning, category, fmt::format("Cannot fetch image information: {}", e.what()));
-            }
+            const auto snapshot = vm.get_snapshot(snapshot_name); // verify validity even if unused
+            if (!snapshot_pick.all_or_none || !snapshots_only)
+                populate_info(vm, snapshot);
         }
+    };
 
-        info->set_image_release(original_release);
-        info->set_id(vm_image.id);
-
-        auto vm_specs = vm_instance_specs[name];
-
-        auto mount_info = info->mutable_mount_info();
-
-        mount_info->set_longest_path_len(0);
+    auto fetch_detailed_report = [&](VirtualMachine& vm) {
+        fmt::memory_buffer errors;
+        const auto& name = vm.vm_name;
 
-        if (!vm_specs.mounts.empty())
-            have_mounts = true;
+        const auto& it = instance_snapshots_map.find(name);
+        const auto& snapshot_pick = it == instance_snapshots_map.end() ? SnapshotPick{{}, true} : it->second;
 
-        if (MP_SETTINGS.get_as<bool>(mp::mounts_key))
+        try
         {
-            for (const auto& mount : vm_specs.mounts)
+            process_snapshot_pick(vm, snapshot_pick);
+            if (snapshot_pick.all_or_none)
             {
-                if (mount.second.source_path.size() > mount_info->longest_path_len())
-                {
-                    mount_info->set_longest_path_len(mount.second.source_path.size());
-                }
-
-                auto entry = mount_info->add_mount_paths();
-                entry->set_source_path(mount.second.source_path);
-                entry->set_target_path(mount.first);
-
-                for (const auto& uid_mapping : mount.second.uid_mappings)
-                {
-                    auto uid_pair = entry->mutable_mount_maps()->add_uid_mappings();
-                    uid_pair->set_host_id(uid_mapping.first);
-                    uid_pair->set_instance_id(uid_mapping.second);
-                }
-                for (const auto& gid_mapping : mount.second.gid_mappings)
-                {
-                    auto gid_pair = entry->mutable_mount_maps()->add_gid_mappings();
-                    gid_pair->set_host_id(gid_mapping.first);
-                    gid_pair->set_instance_id(gid_mapping.second);
-                }
+                if (snapshots_only)
+                    for (const auto& snapshot : vm.view_snapshots())
+                        populate_info(vm, snapshot);
```

Comment:
Again, I think making a way to call `populate_snapshot_info()` here would be better.